src/Stack/Sig/Sign.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE CPP               #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes        #-}
+{-# LANGUAGE TemplateHaskell   #-}
+
+{-|
+Module      : Stack.Sig.Sign
+Description : Signing Packages
+Copyright   : (c) FPComplete.com, 2015
+License     : BSD3
+Maintainer  : Tim Dysinger <tim@fpcomplete.com>
+Stability   : experimental
+Portability : POSIX
+-}
+
+module Stack.Sig.Sign (sign, signTarBytes) where
+
+#if __GLASGOW_HASKELL__ < 710
+import           Control.Applicative ((<$>))
+#endif
+
+import           Network.HTTP.Download
+import           Control.Monad (when)
+import           Control.Monad.Catch
+import           Control.Monad.IO.Class
+import           Control.Monad.Logger
+import           Control.Monad.Trans.Control
+import qualified Data.ByteString.Lazy as L
+import           Data.Functor (void)
+import           Data.List (isSuffixOf)
+import           Data.Monoid ((<>))
+import qualified Data.Text as T
+import           Data.UUID (toString)
+import           Data.UUID.V4 (nextRandom)
+import           Network.HTTP.Conduit
+       (Response(..), RequestBody(..), Request(..), httpLbs, newManager,
+        tlsManagerSettings)
+import           Network.HTTP.Types (status200, methodPut)
+import           Path
+import           Path.IO
+import qualified Stack.Sig.Cabal as Cabal
+import qualified Stack.Sig.GPG as GPG
+import           Stack.Types
+import           System.Directory (getDirectoryContents)
+import           System.Process (readProcessWithExitCode)
+
+-- | Sign a haskell package with the given url of the signature
+-- service and a path to a tarball.
+sign
+    :: (MonadCatch m, MonadBaseControl IO m, MonadIO m, MonadMask m, MonadLogger m, MonadThrow m)
+    => String -> FilePath -> m ()
+sign url filePath = do
```

Comment:
It would definitely be better to use `tar` here, which will avoid the unnecessary filesystem activity.